Regex navigator.

mwfogleman opened the pull request:
For #183: Initial work on a Regex navigator. I have select* working, I think, in Clojure and possibly CLJS. Not sure quite yet how to implement transform*.

Need to:
- [x] Implement transform* for Regexes
- [x] Add select tests
- [x] Add transform tests
- [x] Check and ensure with CLJS
- [x] Update CHANGES.md

mwfogleman commented:
With this patch 

```clojure
=> (select #"a+" "abc")
["a"]
```

nathanmarz commented:
Still isn't right. Should navigate to each match, not to a sequence of matches. Look at the `select*` implementation of `MAP-VALS`.

mwfogleman commented:
Like this?

```clojure
(doseqres NONE [s (re-seq re structure)]
                     (next-fn vals v))
```

nathanmarz commented:
Yes, that looks right

On Wed, Oct 11, 2017 at 10:43 PM, Michael Fogleman <notifications@github.com
> wrote:

> Like this?
>
> (doseqres NONE [s (re-seq re structure)]
>                      (next-fn vals v))
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/nathanmarz/specter/pull/231#issuecomment-336026733>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIBWKT8XszgK2eH5IAHvIc7RRbok_s4ks5sracEgaJpZM4P0tvJ>
> .
>



-- 
Twitter: @nathanmarz
http://nathanmarz.com

mwfogleman commented:
Aha! Yes this is much better.

mwfogleman commented:
I've been working on the transform implementation of the regex navigator. I think I have a better understanding of Java and Clojure's regex handling facilities, but need to integrate that into the Specter implementation. I have a few questions about that.

I was looking at your basic-specter repository to get a sense of how Specter works. Is this still an accurate description (i.e. of 1.0-era Specter?) I noticed that the protocols there do not have the "vals" argument that select* and transform* require. What does "vals" do? (I've tried printing it out in my experiments but, at least with the inputs I've given it, it returns an empty vector.)

I wrote these examples down of what I think we want - the first, select, already works:

```clojure
(= (select #"t" "test") ["t" "t"])
(= (transform #"t" clojure.string/capitalize "test") "TesT")
(= (setval #"t" "z" "test") "zesz")
```

The other thing is, it looks like clojure.string/replace does this already. `(clojure.string/replace structure re next-fn)` should be the value that we want- not the one that is currently in my PR - although I can't return that as the transform* implementation because it does not match the `[path transform-fn structure]` idiom.

My question is, do we want to use that function? If so, how do I supply its return value to Specter? If not, what's a next step? I could see that we might not want to, either because it doesn't match the pattern - so we need to reimplement it (and it's actually built on top of several functions, which we would probably need to reimplement) - or because might not be performant, because it does a few other things, such as replacing one character with another, or one substring with another.

nathanmarz commented:
`vals` is how `collect` / `collect-one` are implemented. You should use `defnav` rather than `defrichnav` since it handles that parameter in the background. `defrichnav` exists since the compiler doesn't do certain inlining optimizations, so `defrichnav` allows the handling of `vals` to be inlined manually. This is only important for navigators which do very little work (like `keypath`) – I think regexes are expensive enough that the inlining optimization won't be noticeable.

Nice discovery on `replace`! This makes the transform implementation an easy `(clojure.string/replace structure pattern next-fn)`Just need to make sure this works in ClojureScript too.

mwfogleman commented:
Awesome! That did the trick! So, transform is now implemented, it seems to work for me in Clojure as expected. I added some tests that work on my end, in Clojure. It also seems that this *should* work in CLJS, as re-seq and clojure.string/replace are available in CLJS. But I'm also getting a failing build for CLJS.

However, I am not able to run the tests currently as it seems my instance of lein is running a JRE without a JDK (even though I have JDK's locally). I'll try and figure this out both for this issue and because I have some other things I'd like to be able to do that require being able to test Specter.

mwfogleman commented:
Alright, I tracked down the CLJS error and tests are passing! This should be functionally done as far as I can tell. I'll work on some internal docs now.

nathanmarz commented:
Two things:

1. Let's rename `regex*` to `regex-nav`. The navigator should also be in specter.cljc. 
2. The tests need to be more extensive. There should be transforms where the transformed result varies based on what is selected (e.g. #"aa*" -> (fn [s] (-> s count str))). There should also be tests of the behavior when using regex capture groups.

mwfogleman commented:
Think that should do it - combining this with subselect is really cool! :)

nathanmarz commented on test/com/rpl/specter/core_test.cljc:
```diff
@@ -1434,6 +1434,19 @@
+  (is (= (transform #"aa*" (fn [s] (-> s count str)) "aadt") "2dt"))
+  (is (= (transform #"[Aa]+" (fn [s] (apply str (take (count s) (repeat "@")))) "Amsterdam Aardvarks") "@msterd@m @@rdv@rks"))
+  (is (= (select [#"(\S+):\ (\d+)" (s/nthpath 2)] "Mary: 1st George: 2nd Arthur: 3rd") ["1" "2" "3"]))
+  (is (= (transform (s/subselect #"\d\w+") reverse "Mary: 1st George: 2nd Arthur: 3rd"))))
```
This expression is incomplete

nathanmarz commented:
Looks like it's almost ready for merge. Just one of the tests needs updating.

mwfogleman commented:
Should be all reformatted now.

nathanmarz commented:
That test I pointed out only has a single argument to `=`

nathanmarz commented:
One last thing, let's take out the implicit nav for regexes for now. I've been limiting the use of implicit navs to objects which can be used as functions. I may reconsider this in the future and make regexes, strings, and numbers have implicit navs.

mwfogleman commented:
This patch should do that. I personally like the implicit functionality for regexes - even though in vanilla Clojure regexes don't work like functions, it made sense to me to treat them the way dropping a keyword into a path in Specter works. In the mean time with this patch the functionality is still accessible and of course anyone can grab the code to use them implicitly.

nathanmarz commented:
Great work, thanks!
